### ISSUE TITLE
Implement font sizing control

### ISSUE THREAD

**Issue by secrettriangle** [daviferreira/medium-editor#376]

**Comment by brunomperes** [comment]
+1

**Comment by sajjad-ser** [comment]
+1

**Comment by hellocosmin** [comment]
+1

**Comment by dmkolobov** [comment]
+1

**Comment by tagisen** [comment]
+1

**Comment by nmielnik** [comment]
Lots of requests for this, we'll be looking into how to do this via button(s) (ie +1 button and -1 button) or via a custom drop-down.  Thanks for the feedback!

**Comment by nmielnik** [comment]
If any of you would like to try to tackle this, I believe you can just add some additional config into `button.js` which would expose 2 buttons.  The action they would need would be `decreaseFontSize` and `increaseFontSize` (or maybe just use `fontSize` for both) as outlined [here](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand)

**Comment by hellocosmin** [comment]
The problem with `decreaseFontSize`, `increaseFontSize` and `fontSize` is that:

1) First two use `<small>` and `<big>` tags, one will most likely need to be more specific and have more granular control over font size values. Also, maybe I need to make a whole paragraph have smaller text: that doesn't mean (semantically speaking) that I should end up with my text inside the `<p>` being wrapped up in `<small>`

2) Second one uses HTML font sizes (1 to 7). Same concern as above.

Maybe un/wrapping the selection in a `<span style="font-size:XXpx">` would suit better? This also makes me think about which measure unit to use (px/em/rem) and if measure unit type should be an option.

Taking that into consideration, I think a custom dropdown, to which you can pass an array of possible font-size values, would be MUCH better than +1/-1 buttons. This way you cover both granular control over font-sizes and the unit type to use.

Thoughts?

**Comment by secrettriangle** [comment]
I have to disagree. I think +1 and -1 buttons, albeit less granular, are more user-friendly. Maybe a slider is a good compromise? I still think +1 and -1 is the way to go. Also, option 2 doesn't sound so bad.

**Comment by hellocosmin** [comment]
A slider sounds awesome, why didn't I think of that? :)

My reasoning for the downvote on buttons is that I presumed it would add 2 more buttons to the toolbar and it can get very long and confusing pretty fast. 

Option 2 might sound nice, but it uses `<font>` which was deprecated in HTML 4.01 and is obsolete in HTML 5. So Option 1 is not semantic, Option 2 is invalid HTML.

I would still vote for a dropdown, but a button that reveals a slider is a much nicer UI&UX touch, I have to agree. The only advantage I can see for a dropdown in this case is that it's just HTML/CSS, while a slider needs additional JS code.

**Comment by nmielnik** [comment]
These are all excellent points.

Some more food for thought here:

* The nice thing about the `<font>` tag is that we can embed CSS that overrides it's behavior.  So you can include css like this:
```css
font[size="3"] {
    font-size: 85%;
}
font[size="4"] {
    font-size: 100%;
}
font[size="5"] {
    font-size: 115%;
}
```
There are some interesting possibilities on how you can customize the font-sizes using that

* The dropdown may still require JS, as any implementation that doesn't involve using `document.execCommand` will need to handle wrapping the selection in a `<span>` with a class, and accounting for any pre-existing classes applied to the selection.  Essentially, we'd have to either mimic, or use what rangy does in its [CSS Class Applier Module](https://github.com/timdown/rangy/wiki/Class-Applier-Module)

**Comment by nmielnik** [comment]
So #494 brings an initial font sizing control into medium-editor (Thanks to @secrettriangle)!.  However, I think we need some collective brainstorming on what we think the behavior should be and how we want to work around some issues with the user's interaction.  

I'm re-opening the issue so that we can smooth out this initial pass of the 'slider' control, as well as determine if we want other variations of this to be available.

@secrettriangle thanks for keeping this fork up to date and I apologize for removing the fontSize stuff from the documentation, but I think we should get a bit more feedback until we make it official and show it off in our demo.

**Comment by nmielnik** [comment]
So unfortunately, IE9 doesn't support sliders, so this current implementation for font-size control won't be sufficient for our browser support as-is.

However, as we've seen before, medium-editor in IE9 doesn't look too great to begin with, so we might want to explore whether we want to continue IE9 support or not.

**Comment by nchase** [comment]
I recently saw Tom Dale and Yehuda Katz of Ember.js put out a community feedback solicitation on a proposal to remove IE8 and IE9 support in Ember 2.0. You can find the discussion [here](https://github.com/emberjs/rfcs/pull/45). It lasted for about ten days and it was really fruitful.

We could write a proposal here as an issue and solicit the same feedback. Maybe we could set feedback expectations and window-for-feedback differently given that this project has smaller scale, audience, and reach than Ember.

I wonder if either of the following are viable options:

* Given that community feedback indicates support for the change, drop _official_ support for IE9 from the next major release. (i.e. stuff may continue to work given the effort put into IE9 so far, but no guarantees, no considerations for IE9 in ongoing development)
* Introduce font-size control, but simply do not support it in IE9.

**Comment by secrettriangle** [comment]
A few good range polyfills:
* https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills#web-forms--individual-input-types

DOM selection polyfills for IE:
* https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills#dom-range-and-selection
* https://github.com/mysticx/IE-Selection-Polyfill
